### PR TITLE
[Docs/Tune] "getting data in and out of Tune" doc

### LINKED ISSUE #28528: [tune] Documentation around syncing / saving / retrieving results is hard to understand
### Description

For the task of "I want to train 1000 models and retrieve each model result later", it's pretty hard to figure out how to do this. There are multiple doc issues:
1. There is no guide for "How to save and retrieve Tune results". This should be a core guide, not something spread out.
2. Several parts of the documentation say "Tune automatically syncs files to the head node". However, this is misleading-- you need to configure syncing, and in many sync modes the files never end up on the head node.
3. There are several ways to do the same thing: tune.log_model, using Checkpoint, saving a file to the current trial dir--- they need to be documented in the same place as different ways to achieve the same goal.
4. There is no documentation on how to retrieve the result of log_model, Checkpoint, etc. after the run has finished.

### Link

_No response_

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

Adds a consolidated guide on getting data in and out of Trainables.

## Related issue number

Closes https://github.com/ray-project/ray/issues/28528

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by justinvyu — COMMENTED** [review]
I like this guide! Left some comments/questions below:

Also, I'm wondering if the last two sections of [this](https://docs.ray.io/en/latest/tune/tutorials/tune-output.html#how-to-configure-trainable-logging) should be referenced somewhere in this guide - this is related to my comment about showing users how to save arbitrary files as a way of "getting data out".

File: doc/source/tune/examples/tune_get_data_in_and_out.ipynb [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+    "        callbacks=[MLflowLoggerCallback(experiment_name=\"example\")]\n",
+    "    ),\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Checkpoints & other artifacts\n",
```
[the remark this answers is not in the document]
Comment (Yard1) [reply doc/source/tune/examples/tune_get_data_in_and_out.ipynb:478]: We do not support that right now. I wanted to leave it open ended enough to be added once we have support for artifact syncing, at which point we can elaborate more.

One simple example is having a log file to which you append.

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 2]
```diff
@@ -0,0 +1,685 @@
...
+First order of business is to provide the inputs for the Trainable. We can broadly separate them into two categories - variables and constants.
+
+Variables are the parameters we want to tune. They will be different for every [Trial](tune_60_seconds_trials). For example, those may be the learning rate and batch size for a neural network, number of trees and the maximum depth for a random forest, or the data partition if you are using Tune as an execution engine for batch training.
+
+Constants are the parameters that are the same for every Trial. Those can be the number of epochs, model hyperparameters we want to set but not tune, the dataset and so on. Often, the constants will be quite large (eg. the dataset or the model).
+
+### Search space
+
+```{note}
+TL;DR - use the `param_space` argument to specify small, serializable parameters that you want to tune over.
```
Comment (justinvyu) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:58]:
```suggestion
TL;DR - use the `param_space` argument to specify small, serializable constants and variables.
```

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 3]
```diff
@@ -0,0 +1,685 @@
...
+    for epoch in range(epochs):
+        # Simulate doing something expensive.
+        time.sleep(1)
+        metric = (0.1 + model["hyperparameter_a"] * epoch / 100) ** (
+            -1
+        ) + model["hyperparameter_b"] * 0.1 * data["A"].sum()
+        trained_model = {"state": model, "epoch": epoch}
+
+
+tuner = Tuner(
```
Comment (justinvyu) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:134]: We should move this tuner with param space to the previous section to show how to specify `param_space`.

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+        "hyperparameter_b": tune.uniform(-100, 100),
+        "epochs": 10,
+    },
+    tune_config=tune.TuneConfig(num_samples=4),
+)
+```
+
+### Loading data in Trainable
+
+You can also load data directly in Trainable from eg. cloud storage, NFS or from disk.
```
Comment (justinvyu) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:163]:
```suggestion
You can also load data directly in Trainable from eg. cloud storage, shared file storage such as NFS, or from the local disk of the Trainable worker.
```

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 5]
```diff
@@ -0,0 +1,685 @@
...
+
+```{warning}
+When loading from disk, ensure that all nodes in your cluster have access to the file you are trying to load.
+```
+
+A common use-case is to load the dataset from S3 or any other cloud storage with pandas, arrow or any other framework.
+
+The working directory of the Trainable worker will be automatically changed to the corresponding Trial directory. For more details, see {ref}`tune-working-dir`.
+
+For brevity, we will not use this approach in our example.
```
Comment (justinvyu) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:173]:
```suggestion
```

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 6]
```diff
@@ -0,0 +1,685 @@
...
+### Reporting metrics
+
+*Metrics* are values passed through the `metrics` argument in a `session.report` call. Metrics can be used by Tune [Search Algorithms](search-alg-ref) and [Schedulers](schedulers-ref) to direct the search. After the tuning run is complete, you can [analyze the results](/tune/examples/tune_analyze_results), which include the reported metrics.
+
+```{warning}
+Similarly to search space values, each value reported as a metric will be saved directly in the Trial metadata. This means that every value reported as a metric **must** be serializable and take up a small amount of memory.
+```
+
+```{note}
+Tune will automatically include some metrics, such as the training iteration, timestamp and more.
```
Comment (justinvyu) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:278]:
```suggestion
Tune will automatically include some metrics, such as the training iteration, timestamp and more. See [here](tune-autofilled-metrics) the entire list.
```

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 7]
```diff
@@ -0,0 +1,685 @@
...
+def training_function(config, data):
+    model = {
+        "hyperparameter_a": config["hyperparameter_a"],
+        "hyperparameter_b": config["hyperparameter_b"],
+    }
+    epochs = config["epochs"]
+
+    # Load the checkpoint, if there is any.
+    loaded_checkpoint = session.get_checkpoint()
+    if loaded_checkpoint is not None:
```
Comment (justinvyu) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:393]: Does this need `loaded_checkpoint.to_dict()`?

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 8]
```diff
@@ -0,0 +1,685 @@
...
+With all of those changes implemented, we can now run our tuning and obtain meaningful metrics and artifacts.
+
+
+```python
+results = tuner.fit()
+results.get_dataframe()
+```
+
+
+<div class="tuneStatus">
```
Comment (justinvyu) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:435]: Can we remove this tuner output and only show the dataframe? The table shows similar information and is not formatted that well in the markdown.

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 9]
```diff
@@ -0,0 +1,685 @@
...
+For brevity, we will not use this approach in our example.
+
+Our tuning run can now be run, though we will not yet obtain any meaningful outputs back.
+
+
+```python
+results = tuner.fit()
+```
+
+    2022-11-30 17:39:53,677	INFO worker.py:1529 -- Started a local Ray instance. View the dashboard at [1m[32mhttp://127.0.0.1:8265 [39m[22m
```
Comment (justinvyu) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:182]: Maybe remove the tuner outputs here as well.

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+      <td>/home/ubuntu/ray_results/training_function_202...</td>
+    </tr>
+  </tbody>
+</table>
+<p>4 rows × 23 columns</p>
+</div>
+
+
+
+Checkpoints, metrics, and the log directory for each trial can be accessed through the `ResultGrid` output of a Tune experiment. For more information on how to interact with the returned `ResultGrid`, see {doc}`/tune/examples/tune_analyze_results`.
```
Comment (ericl) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:516]: How do you access this after the job/cluster has been terminated? Actually this was the original user question behind this ticket.

**Review by ericl — COMMENTED** [review]
I enjoyed reading this--- a few comments and a missing section I think.

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+
+Constants are the parameters that are the same for every Trial. Those can be the number of epochs, model hyperparameters we want to set but not tune, the dataset and so on. Often, the constants will be quite large (eg. the dataset or the model).
+
+### Search space
+
+```{note}
+TL;DR - use the `param_space` argument to specify small, serializable constants and variables.
+```
+
+The first way of passing inputs into Trainables is the [*search space*](tune-key-concepts-search-spaces) (it may also be called *parameter space* or *config*). In the Trainable itself, it maps to the `config` dict passed in as an argument to the function. You define the search space using the `param_space` argument of the `Tuner`. The search space is a dict and may be composed of [*distributions*](<tune-sample-docs>), which will sample a different value for each Trial, or out of constant values. The search space may be composed out of nested dictionaries, and those in turn can have distributions as well.
```
Comment (krfricke) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:61]:
```suggestion
The first way of passing inputs into Trainables is the [*search space*](tune-key-concepts-search-spaces) (it may also be called *parameter space* or *config*). In the Trainable itself, it maps to the `config` dict passed in as an argument to the function. You define the search space using the `param_space` argument of the `Tuner`. The search space is a dict and may be composed of [*distributions*](<tune-sample-docs>), which will sample a different value for each Trial, or of constant values. The search space may be composed of nested dictionaries, and those in turn can have distributions as well.
```

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+tuner = Tuner(training_function, tune_config=tune.TuneConfig(num_samples=4))
+```
+
+## Getting data in
+
+First order of business is to provide the inputs for the Trainable. We can broadly separate them into two categories - variables and constants.
+
+Variables are the parameters we want to tune. They will be different for every [Trial](tune_60_seconds_trials). For example, those may be the learning rate and batch size for a neural network, number of trees and the maximum depth for a random forest, or the data partition if you are using Tune as an execution engine for batch training.
+
+Constants are the parameters that are the same for every Trial. Those can be the number of epochs, model hyperparameters we want to set but not tune, the dataset and so on. Often, the constants will be quite large (eg. the dataset or the model).
```
Comment (krfricke) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:53]:
```suggestion
Constants are the parameters that are the same for every Trial. Those can be the number of epochs, model hyperparameters we want to set but not tune, the dataset and so on. Often, the constants will be quite large (e.g. the dataset or the model).
```

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+        "hyperparameter_b": tune.uniform(-100, 100),
+        "epochs": 10,
+    },
+    tune_config=tune.TuneConfig(num_samples=4),
+)
+```
+
+### Loading data in Trainable
+
+You can also load data directly in Trainable from eg. cloud storage, shared file storage such as NFS, or from the local disk of the Trainable worker.
```
Comment (krfricke) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:173]:
```suggestion
You can also load data directly in Trainable from e.g. cloud storage, shared file storage such as NFS, or from the local disk of the Trainable worker.
```

**Review by krfricke — COMMENTED** [review]
This is great, couple of suggestions

**Comment by Yard1** [comment]
@ericl Added a section on accessing results later, PTAL

File: doc/source/tune/tutorials/tune_get_data_in_and_out.md [hunk 14]
```diff
@@ -514,3 +518,10 @@ results.get_dataframe()
 
 
 Checkpoints, metrics, and the log directory for each trial can be accessed through the `ResultGrid` output of a Tune experiment. For more information on how to interact with the returned `ResultGrid`, see {doc}`/tune/examples/tune_analyze_results`.
+
+
+### How do I access results after I am finished?
+
+After you have finished running the Python session, you can still access the results and checkpoints. By default, Tune will save the experiment results to the `~/ray_results` local directory. You can configure Tune to persist results in the cloud as well. See {ref}`tune-checkpoint-syncing` for more information on how to configure syncing.
+
+You can restore the Tune experiment by calling `Tuner.restore(path_or_cloud_uri)`, where `path_or_cloud_uri` points to a location either on the filesystem or cloud where the experiment was saved to. After the `Tuner` has been restored, you can access the results and checkpoints as outlined in the previous section.
```
Comment (ericl) [inline doc/source/tune/tutorials/tune_get_data_in_and_out.md:527]: Mention you can use `tuner.get_results()` to get the previous results without fitting.

**Review by ericl — APPROVED** [review]
Just a couple nits.
